src/server/server_family.cc

```diff
@@ -1629,11 +1631,65 @@ void ServerFamily::Memory(CmdArgList args, ConnectionContext* cntx) {
   return mem_cmd.Run(args);
 }
 
+void ServerFamily::BgSaveFb(bool new_version, std::string basename,
+                            boost::intrusive_ptr<Transaction> trans) {
+  GenericError ec = DoSave(new_version, basename, trans.get());
+  if (ec) {
+    LOG(INFO) << "Error in BgSaveFb: " << ec.Format();
+  }
+  {
+    lock_guard lk(bg_save_mu_);
+    bg_save_fb_done_ = true;
+  }
+}
+
+// BGSAVE [DF|RDB] [basename]
+// TODO add missing [SCHEDULE]
+void ServerFamily::BgSave(CmdArgList args, ConnectionContext* cntx) {
+  if (args.size() > 2) {
+    return cntx->SendError(kSyntaxErr);
+  }
+
+  bool new_version = absl::GetFlag(FLAGS_df_snapshot_format);
+
+  if (args.size() >= 1) {
+    ToUpper(&args[0]);
+    string_view sub_cmd = ArgS(args, 0);
+    if (sub_cmd == "DF") {
+      new_version = true;
+    } else if (sub_cmd == "RDB") {
+      new_version = false;
+    } else {
+      return cntx->SendError(UnknownSubCmd(sub_cmd, "SAVE"), kSyntaxErrType);
+    }
+  }
+
+  string_view basename;
+  if (args.size() == 2) {
+    basename = ArgS(args, 1);
+  }
+
+  {
+    lock_guard lk(bg_save_mu_);
+
+    if (bg_save_fb_.IsJoinable() && !bg_save_fb_done_) {
+      return cntx->SendError("BgSave in process");
+    }
+    if (bg_save_fb_done_) {
+      // non blocking, fiber is in Terminate state
+      bg_save_fb_.Join();
+    }
+    bg_save_fb_done_ = false;
+    bg_save_fb_ = fb2::Fiber("bg_save_fiber", &ServerFamily::BgSaveFb, this, new_version,
+                             std::string(basename), cntx->transaction);
+  }
+  cntx->SendOk();
+}
```

Comment:
We already have save exclusivity with `save_mu_`, why do we want to manage bgsave separately? We can store no information at all about it